File: code/12-JavaCodeLoop/JavaTDDLoop/src/main/java/com/example/valueobjects/GameResult.java
package com.example.valueobjects;

/**
 * The outcome of a single {@code GuessGame} play-through.
 */
public record GameResult(int secretNumber, int attempts, boolean success) {

    public GameResult {
        if (attempts <= 0) {
            throw new IllegalArgumentException("attempts must be positive, was " + attempts);
        }
    }

    public static GameResult success(int secretNumber, int attempts) {
        return new GameResult(secretNumber, attempts, true);
    }

    public static GameResult failure(int secretNumber, int attempts) {
        return new GameResult(secretNumber, attempts, false);
    }
}

File: code/13-Java2Python/python/valueobjects/game_result.py
from dataclasses import dataclass


@dataclass(frozen=True)
class GameResult:
    """The outcome of a single GuessGame play-through."""

    secret_number: int
    attempts: int
    success: bool

    def __post_init__(self) -> None:
        if self.attempts <= 0:
            raise ValueError(f"attempts must be positive, was {self.attempts}")

    @staticmethod
    def won(secret_number: int, attempts: int) -> "GameResult":
        return GameResult(secret_number, attempts, True)

    @staticmethod
    def lost(secret_number: int, attempts: int) -> "GameResult":
        return GameResult(secret_number, attempts, False)
